### ISSUE TITLE
PVI behavior

### ISSUE THREAD

**Issue by mbronckers** [mbronckers/GI-PVI#5]
This thread serves to log any observations regarding PVI behavior, including homogenous/heterogeneous data partitions and single/multi-client PVI.

**Comment by mbronckers** [comment]
The `split_data_clients()` creates a random permutation based on the list of splits provided. In the case that only one client is selected, it still generates a random permutation of the training dataset. This means that the results do not exactly line up with the previous commit anymore.

Still, the observed behavior is in my opinion conforming to our expectation, i.e. appropriately fitting to the training data and having uncertainty outside of it.

Left is no random permutation of the training dataset, right is with random permutation (seed 0).
<img width="1666" alt="image" src="https://user-images.githubusercontent.com/25270093/174402609-63e30386-7ac5-4008-9240-f50fd77e8de4.png">

**Comment by mbronckers** [comment]
- Fixing the global VI file (regression.py) by giving each client a `vs` object led to backward() errors (i.e. need to retain graph). Odd. Maybe that's why we face the same issue in PVI. 

- Good news: after updating the client_nz , the PVI behavior looks to find the global VI solution. 
- in the single client case, we expect exact retrieval. It looks more tightly fitting through its own data points. This might be attributable to the number of iters&epochs (10, 100 respectively), but the global VI was also run for 1000 epochs.

**Comment by mbronckers** [comment]
- single-client PVI, 3 client PVI, and global VI (left-to-right):
<img width="1951" alt="image" src="https://user-images.githubusercontent.com/25270093/175663528-02132e14-f874-4f86-bd64-155fe5e07d89.png">

**Comment by mbronckers** [comment]
- Detaching the error predictions in PVI resolved the backward()/retain_graph issue.

**Comment by mbronckers** [comment]
> - There is a backward()/retain_graph issue in global VI. Makes it very slow. Fix TBA. 
- Global VI backward()/retain_graph issue is fixed. Had to do with updating client's nz.

> - Single client PVI with 1000 epoch & 1 iter should be equal to 100 epoch & 10 iters , right?
> - PVI should recover global VI solution. How close should it be?

Now option to fix ll var in global VI.

**Comment by mbronckers** [comment]
Great news: **Single-client sequential PVI recovers identical solution (also intermediate steps) as global VI** meeting expectations.

<img width="1585" alt="image" src="https://user-images.githubusercontent.com/25270093/175829868-a5aa9bd5-3431-4b3f-8109-9875c3c8e3b6.png">

**Comment by mbronckers** [comment]
Interesting debugging finding in GVI, not sure of the implications:
- In the final layer, the mean of the client's likelihood factor cannot be computed because the factor's precision is not positive-definite [`B.chol(self.prec)` throws an error]. 
- However, because q has the prior's precision (1 on the diagonal) added to the client's factor, q's precision is positive-definite and the mean can be computed.

**Comment by mbronckers** [comment]
Ober's inducing point behavior when (z, yz) are initialized to (X_train, y_train):
<img width="498" alt="image" src="https://user-images.githubusercontent.com/25270093/176512551-dfd83bca-639e-46e2-ae98-223b5c35e373.png">

and the predictive distribution using a StandardPrior:
![image](https://user-images.githubusercontent.com/25270093/176512780-e4cc0559-b79e-49fd-898c-63798bd034ca.png)

and the predictive distribution using a NealPrior:
<img width="480" alt="image" src="https://user-images.githubusercontent.com/25270093/176512705-ed6f8be1-e124-47dd-99b2-5a78992e6447.png">

**Comment by mbronckers** [comment]
One thing that really differs is that some of our inducing points do move towards the center region, in certain parameter configurations. I am starting to believe this has something to do with their effectively scaled learning rates, but will run our own with different learning rates

The other aspect I am considering is their usage of the prior.
- Their results are only good when they use their NealPrior
- For us the NealPrior results in a too tight distribution that barely fits to the data.
- But when I run a HePrior (instead of NealPrior which sets variance to 1/dims_in, set variance to 2/dims_in), our results look really good and in line with theirs.

I haven’t found a reason why our NealPrior doesn’t work as well as theirs but a He does.

**Comment by mbronckers** [comment]
The KLs are approximately equal in the first epoch. Minor differences are attributable to (1) logpq computation vs q.kl(p_); (2) debug sessions run on different devices => same seed, but X, y can be slightly different; (3) not sure if bayesfunc specifies the seed in every stochastic update.
![image](https://user-images.githubusercontent.com/25270093/176536548-d5c22f72-e7de-464f-aa1a-21c0f3979941.png)

![image](https://user-images.githubusercontent.com/25270093/176536689-16719335-5c08-4abc-8584-e14f439c0647.png)

**Comment by mbronckers** [comment]
- Ober's likelihood computation is different. They take the mean across N data points (leaving an average ll per inference sample?) and then subtract a KL (logpq) but divided by the batch size (len(x)). 

- Dividing the KL by the batch size for us gives nice results. It effectively dilutes the KL term effect on ELBO and prevents the inducing points from going towards the prior too much

**Comment by mbronckers** [comment]
Ober's inducing VP behavior: 
<img width="802" alt="image" src="https://user-images.githubusercontent.com/25270093/176646367-3fbb49a4-fdb3-4d05-9a12-f7c083bee1a0.png">

Ours:
<img width="835" alt="image" src="https://user-images.githubusercontent.com/25270093/176648792-153f6183-c452-42ba-bf4e-4b918e301171.png">

**Comment by mbronckers** [comment]
Ober's predictive distribution:
<img width="646" alt="image" src="https://user-images.githubusercontent.com/25270093/176649382-e316cffe-966e-412b-be51-d88d24783cdd.png">
<img width="644" alt="image" src="https://user-images.githubusercontent.com/25270093/176649404-cdb43def-4884-43df-b24e-3eabd000721f.png">

Ours:
<img width="836" alt="image" src="https://user-images.githubusercontent.com/25270093/176649511-87f16248-173d-426e-9c6d-35bdf3086af4.png">
<img width="829" alt="image" src="https://user-images.githubusercontent.com/25270093/176649880-0f5216f7-e29d-43fd-b336-585fa5ebfb22.png">

**Comment by mbronckers** [comment]
Our final layer nz:
<img width="643" alt="image" src="https://user-images.githubusercontent.com/25270093/176655979-7db26671-130a-48b3-bce4-475993920bf1.png">

Ober final layer nz:
<img width="603" alt="image" src="https://user-images.githubusercontent.com/25270093/176656026-2ab491a5-a98a-4181-97db-ea309d12ae67.png">

**Comment by mbronckers** [comment]
Ober middle layer nz:
<img width="618" alt="image" src="https://user-images.githubusercontent.com/25270093/176656434-8c9022e5-2bbe-40fa-b57c-24d1401759e8.png">

Our middle layer nz:
<img width="555" alt="image" src="https://user-images.githubusercontent.com/25270093/176656509-556848b0-e425-4c47-8a9a-81d2d00143f3.png">

**Comment by mbronckers** [comment]
Ober first layer nz:
<img width="616" alt="image" src="https://user-images.githubusercontent.com/25270093/176656974-6d24b44d-078a-4c0d-95e8-cf93db6a7b24.png">

Our first layer nz:
<img width="584" alt="image" src="https://user-images.githubusercontent.com/25270093/176657000-ac058215-82b9-4a04-816b-121b00d29eeb.png">

**Comment by mbronckers** [comment]
Ober 
<img width="860" alt="image" src="https://user-images.githubusercontent.com/25270093/176658554-cf598e62-e77d-4d02-801b-2d78f99b5323.png">

<img width="914" alt="image" src="https://user-images.githubusercontent.com/25270093/176662205-c08ccd0f-a99b-430a-932f-4dc14c28d3ee.png">

**Comment by mbronckers** [comment]
**Good news**: it appears that my single-client PVI implementation recovers the same results as Ober's GI method, scaling the KL in the ELBO computation by 1/batch_size.

Some observations:
- In their paper, they say they initialize the final layer's precision to 1, but in their default layer settings in their codebase they initialize it to the same as other layers (exp(-4)). 

Initializing the final layer's precision to (_left_: exp(-4), _right_: 1) results in the predictive distribution below: 
<img width="2054" alt="image" src="https://user-images.githubusercontent.com/25270093/176676859-d74ca16a-90a3-422c-b0d6-9ccdbfad6b25.png">
<img width="2056" alt="image" src="https://user-images.githubusercontent.com/25270093/176675709-9a638d26-a09b-4147-ba54-0f3e7c91d10c.png">.
and the following inducing points:
<img width="2052" alt="image" src="https://user-images.githubusercontent.com/25270093/176675846-bdc1873c-5202-49d7-b4cc-e78b27278908.png">

Resulting differences:
- slightly more curved predictive distribution
- slightly more uncertainty outside of the domain, when init yz prec to 1. 
- different inducing point behavior
- The inducing point behavior of exp(-4) is the same as Ober's inducing point behavior (see post above). But their predictive distribution is more curved like our prec_yz 1 initialization
-

**Comment by mbronckers** [comment]
# Identical results obtained

Finally, have obtained identical results to Ober.

The final step was to set `neuron_prec` to True in their GI layers, which creates a precision matrix of shape [Dout x 1 x N] . Their code says “**neuron_prec:** Use a different precision parameter for each hidden neuron?  Considerably increases computational cost for relatively small performance benefit.  Default: ``False``.”
![image](https://user-images.githubusercontent.com/25270093/177036343-da032a8a-b9bf-4b29-81be-6d7bc39b71c6.png)

**Results**:
![image](https://user-images.githubusercontent.com/25270093/177036295-ad1311a4-9e35-4a3b-adf2-6063d109bb6b.png)
![image](https://user-images.githubusercontent.com/25270093/177036306-f1d7536b-11ea-473c-bc1d-6c0f2158b33d.png)

<img width="1672" alt="image" src="https://user-images.githubusercontent.com/25270093/177036273-4419873c-d224-4c1c-90c1-72e1643e118c.png">

**Comment by mbronckers** [comment]
The above also relies on using the exact same computational backend functions (`torch.triangular_solve` instead of `solve_triangular`) because of numerical differences (see #8)

![image](https://user-images.githubusercontent.com/25270093/177036387-a09c0dff-8b01-418f-8357-da9efec71591.png)

**Comment by mbronckers** [comment]
## Comparison of the gradients on the first epoch

**inducing points**:
![image](https://user-images.githubusercontent.com/25270093/177036900-31769a19-d26e-4809-a898-ce9f460e012e.png)

**yz layer 1**:
<img width="711" alt="image" **src="https://user-images.githubusercontent.com/25270093/177036888-def92553-1665-47ac-b450-91f23d9f36b5.png">**

**nz layer 1**:
<img width="713" alt="image" src="https://user-images.githubusercontent.com/25270093/177036907-27365064-c866-490f-aeb9-8c65069d7415.png">

**yz layer 2**:
<img width="712" alt="image" src="https://user-images.githubusercontent.com/25270093/177036918-3eaee272-799d-4b40-a664-10cc7c184f0f.png">

**nz layer 2**
<img width="709" alt="image" src="https://user-images.githubusercontent.com/25270093/177036931-89666f69-b0e0-45c6-a3c2-f75b42728fc3.png">

**yz layer 3 (pseudo-obs)**
<img width="708" alt="image" src="https://user-images.githubusercontent.com/25270093/177036947-79e227c6-acd0-4937-bdfe-69af799dce40.png">

**nz layer 3**:
<img width="713" alt="image" src="https://user-images.githubusercontent.com/25270093/177036956-c91f660a-f658-4f65-b8a7-4d5891596eee.png">

**Comment by mbronckers** [comment]
Funnily enough, when we switch back to `B.triangular_solve` (= `torch.triangular_solve`), we blow up our KL and results. So weird. 
```
dW, _ = torch.triangular_solve(noise, B.dense(B.chol(self.prec)), upper=False, transpose=True)
# vs
dW = B.triangular_solve(B.dense(B.chol(self.prec).T), noise, lower_a=True)  # Us
```

<img width="1880" alt="image" src="https://user-images.githubusercontent.com/25270093/177038012-b1028e2e-6fd5-48bd-9756-6e87900db0b4.png">
